### PR TITLE
feat(subtypes): Add grpc interceptor for transforming attributes

### PR DESCRIPTION
**Pull request by tmessi**
This interceptor is responsible for maintaining compatability with the JSON
API with the change to the protobuf messages. The messages are changing
from a single attribue field to a oneof to explicitly associate subtype
attribute messages with the resource message. The interceptor transforms
the request prior to the auditReqeust interceptor, and transforms the
response after the auditResponse interceptor. This allows the audit
interceptors to use all the necessary classification information when
creating the audit events.

### CODE REVIEW

File: internal/types/subtypes/interceptor.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor.go:37]: Could we add a comment here to explain the assumptions that underlay these? E.g.

```
* "item", if it exists, is assumed to be a field in the top level request containing the primary model of the service, e.g. an Account used in a Create request.
* "id", if it exists, is assumed to be a field in the model containing the models id, from which we can infer the subtype of the request.
* "attributes" is assumed to be a field in the model containing a `google.protobuf.Struct` with fields mapping to one of the specific subtypes, or a generic attribute structure if the type is not specific.
```

File: internal/types/subtypes/interceptor.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor.go:48]: Could we add a comment explaining the scenario that this implies and the next steps? E.g.
```
If "item" is set, it implies that we're sending a resource in the request, e.g. via a Create call.
Use the following priority for deducing the subtype:
* Look it up from the "id" field, if set.
* Look it up from any specified source ID field.
* Look it up by from the "type" field,
```

I think this code is pretty self documenting, but I'm worried it's because we've spent weeks in reflect land already, and whenever reflection is concerned, it's never a bad idea to over-comment, IMO.

File: internal/types/subtypes/interceptor.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor.go:76]: Should we surface the error somehow? Goes for all instances of this call.

File: internal/types/subtypes/interceptor.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor.go:77]: Same here re: commenting

File: internal/types/subtypes/interceptor.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor.go:106]: Could we add a comment for the scenarios we expect here similarly to the request?

File: internal/types/subtypes/interceptor.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor.go:129]: Nit: remove this newline?

File: internal/types/subtypes/interceptor.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor.go:178]: Do we need the context parameter?

File: internal/types/subtypes/interceptor_ext_test.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor_ext_test.go:19]: Can we name this `handlerResp` to disambiguate from `handlerResult`? I initially read "result" not "response".

File: internal/types/subtypes/interceptor_ext_test.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor_ext_test.go:21]: Obviously this becomes `expectResp` too.

File: internal/types/subtypes/source_registry.go
Comment (johanbrandhorst) [inline internal/types/subtypes/source_registry.go:53]: Is this safe to assert? What if this option is not set on the field?

File: internal/types/subtypes/attribute_transform.go
Comment (johanbrandhorst) [inline internal/types/subtypes/attribute_transform.go:125]: Add comments to underscore all the assumptions about field names and structure, please.

File: internal/types/subtypes/attribute_transform.go
Comment (johanbrandhorst) [inline internal/types/subtypes/attribute_transform.go:159]: Worth commenting that this also unset the old field, by virtue of `defaultAttrField` being part of a `oneof`.

File: internal/types/subtypes/attribute_transform.go
Comment (johanbrandhorst) [inline internal/types/subtypes/attribute_transform.go:121]: 😬

File: internal/types/subtypes/attribute_transform.go
Comment (tmessi) [reply internal/types/subtypes/attribute_transform.go:158]: That is documents on [Set](https://github.com/protocolbuffers/protobuf-go/blob/32051b4f86e54c2142c7c05362c6e96ae3454a1c/reflect/protoreflect/value.go#L95-L96) do you think it is needed here as well? We call set on a one of in a couple spots in this file. Would you want a comment in both places?

File: internal/types/subtypes/interceptor.go
Comment (tmessi) [reply internal/types/subtypes/interceptor.go:178]: The other [interceptors](https://github.com/hashicorp/boundary/blob/tmessi-subtype-grpc-interceptor/internal/servers/controller/gateway.go) all take a context. I think it is likely that this context will be needed or used at some point, since we need a context to make error events. So I think it should be included even if it is not used yet.

File: internal/types/subtypes/interceptor.go
Comment (tmessi) [reply internal/types/subtypes/interceptor.go:25]: This does seem to be safe, and will result in an empty string being returned if domain is not set on the file. I don't think an error is necessary, since the only thing I can see to do if a domain is set, is potentially return from `transformRequestAttributes` if `domain == ""`

File: internal/types/subtypes/attribute_transform.go
Comment (tmessi) [reply internal/types/subtypes/attribute_transform.go:95]: Ah, yep. I forgot I left this. I should address this in this PR.

File: internal/types/subtypes/attribute_transform.go
Comment (tmessi) [reply internal/types/subtypes/attribute_transform.go:77]: I am not entirely sure, these `structToProto` and `protoToStruct` functions are lifted from the [handlers pkg](https://github.com/hashicorp/boundary/blob/tmessi-subtype-grpc-interceptor/internal/servers/controller/handlers/subtypes.go). I think we should be able to remove the functions from `handlers` once we have updated all of the packages. Then this functionality should remain internal to the `subtypes` package.

File: internal/types/subtypes/attributes.go
Comment (tmessi) [reply internal/types/subtypes/attributes.go]: Yea this gets fixed in 026f8b730, I think this will get squashed together before merging, since this is tightly coupled with the grpc interceptor. I think I just missed adding the correct doc in this earlier commit.

File: internal/types/subtypes/registry_global.go
Comment (tmessi) [reply internal/types/subtypes/registry_global.go:89]: Good catch, I just need to thread a context.Context into this to avoid using `errors.NewDeprecated`.

File: internal/types/subtypes/attribute_transform.go
Comment (tmessi) [reply internal/types/subtypes/attribute_transform.go:101]: this will look different once I address the TODO mentioned [here](https://github.com/hashicorp/boundary/pull/1967#discussion_r842048065)

File: internal/types/subtypes/attribute_transform.go
Comment (johanbrandhorst) [reply internal/types/subtypes/attribute_transform.go:158]: I think it would be nice, yeah.

File: internal/types/subtypes/interceptor.go
Comment (tmessi) [reply internal/types/subtypes/interceptor.go:177]: They are not scoped to the request context. If you look at the other interceptors, most take a context but do not use it, the ones that do use it are performing [initialization checks](https://github.com/hashicorp/boundary/blob/dd988c1d867c813759a5aa0f2fb1151c69791cf6/internal/servers/controller/interceptor.go#L55-L73). I don't know what initialization checks could be added to this interceptor, but I think consistency with the other interceptors is a good idea here.

Comment (johanbrandhorst) [reply internal/types/subtypes/interceptor.go:177]: That sounds reasonable 👍🏻, thanks for explaining.

File: internal/types/subtypes/interceptor.go
Comment (tmessi) [reply internal/types/subtypes/interceptor.go:48]: I can look to add some additional comments in this file. I went back and forth on this a bit. I agree that it is a bit "self documenting", and I think the corresponding tests also provide some good context. The only down-side I have see to over-commenting is when the code gets updated and the comments dont.

Comment (johanbrandhorst) [reply internal/types/subtypes/interceptor.go:48]: Yeah, agree, hopefully this code won't change too much

File: internal/types/subtypes/interceptor.go
Comment (tmessi) [reply internal/types/subtypes/interceptor.go:102]: I added docs in [7658e9e](https://github.com/hashicorp/boundary/pull/1967/commits/7658e9e489454006241899d5864127cb99e47422) that should hopefully explain this now

File: internal/types/subtypes/interceptor.go
Comment (tmessi) [reply internal/types/subtypes/interceptor.go:113]: See [7658e9e](https://github.com/hashicorp/boundary/pull/1967/commits/7658e9e489454006241899d5864127cb99e47422)

File: internal/types/subtypes/interceptor.go
Comment (tmessi) [reply internal/types/subtypes/interceptor.go:201]: See [7658e9e](https://github.com/hashicorp/boundary/pull/1967/commits/7658e9e489454006241899d5864127cb99e47422)

File: internal/types/subtypes/attribute_transform.go
Comment (tmessi) [reply internal/types/subtypes/attribute_transform.go:140]: See [24fe7b8](https://github.com/hashicorp/boundary/pull/1967/commits/24fe7b8a7acdd5176c291e72452144c3005a260e)

File: internal/types/subtypes/attribute_transform.go
Comment (tmessi) [reply internal/types/subtypes/attribute_transform.go:97]: Actually. I think we can revisit this when we get to `authmethods` that is the only package (currently) that requires this functionality. But the general approach will be to create another proto custom option that can be set on the message that will indicate if `DiscardUnknown` should be set to true.

Comment (johanbrandhorst) [reply internal/types/subtypes/attribute_transform.go:97]: OK

File: internal/types/subtypes/attribute_transform.go
Comment (johanbrandhorst) [inline internal/types/subtypes/attribute_transform.go:18]: This comment is a little confusing, why, when there is an error, is there no need to convert?

File: internal/types/subtypes/attribute_transform.go
Comment (johanbrandhorst) [inline internal/types/subtypes/attribute_transform.go:162]: Would it be worth adding a comment here about what we're doing? The only clue is the `if` statement, so maybe worth spelling it out.
```suggestion
	// Find the populated oneof field and turn it into a `*structpb.Struct`.
	for i := 0; i < oneOfFields.Len(); i++ {
```

File: internal/types/subtypes/attributes.go
Comment (johanbrandhorst) [inline internal/types/subtypes/attributes.go:47]:
```suggestion
// does not provide a field with a subtype of "default", or if the default
// field is not a *structpb.Struct type, an error is returned.
```

File: internal/types/subtypes/interceptor_ext_test.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor_ext_test.go:294]: Could we add a test here as well to see what happens when the response contains an unknown `type`?

File: internal/types/subtypes/interceptor_test.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor_test.go:575]: Should we have an error test as well for these functions? Personally I'd probably argue that we don't necessarily need both the internal and external tests so I'll leave it up to you.

File: internal/types/subtypes/registry_global.go
Comment (johanbrandhorst) [inline internal/types/subtypes/registry_global.go:95]: Nice 👌🏻

File: internal/types/subtypes/interceptor.go
Comment (johanbrandhorst) [inline internal/types/subtypes/interceptor.go:95]: This comment is ⭐

**Review by johanbrandhorst — COMMENTED** [review]
Minor doc and test nits remaining

File: internal/types/subtypes/attribute_transform.go
Comment (tmessi) [reply internal/types/subtypes/attribute_transform.go:18]: I can expand the comment to clarify. But `attribueField` will return an error if the given message is not registered, which will be the case for a message like [this test case](https://github.com/hashicorp/boundary/blob/fc2447052a6ddc85da8be422437f13675f5878aa/internal/types/subtypes/interceptor_test.go#L247-L271). So if we are not able to get a `defaultAttrField` there should be no need to convert.

Comment (johanbrandhorst) [reply internal/types/subtypes/attribute_transform.go:18]: Yeah I think it would be valuable to expand the comment 👍🏻.

File: internal/types/subtypes/interceptor_test.go
Comment (tmessi) [reply internal/types/subtypes/interceptor_test.go:575]: I think in this case the external test is more important, since the particular error message and type is important as it will be returned. And that error message is determined in the exported interceptor function. The particular errors from the un-exported functions imo should be free to change.

Comment (johanbrandhorst) [reply internal/types/subtypes/interceptor_test.go:575]: Agreed the external tests are more important in general.

File: internal/types/subtypes/interceptor_ext_test.go
Comment (tmessi) [reply internal/types/subtypes/interceptor_ext_test.go:316]: See d6d2e1f38

File: internal/types/subtypes/attribute_transform.go
Comment (tmessi) [reply internal/types/subtypes/attribute_transform.go]: See 52cf23ce7

**Review by johanbrandhorst — APPROVED** [review]

**Pull request merged**
